staging/src/k8s.io/apiserver/pkg/util/peerproxy/peerproxy_handler.go

```diff
@@ -0,0 +1,369 @@
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		ctx := r.Context()
+		requestInfo, ok := apirequest.RequestInfoFrom(ctx)
+
+		if !ok {
+			responsewriters.InternalError(w, r, errors.New("no RequestInfo found in the context"))
+			return
+		}
+
+		// Allow non-resource requests
+		if !requestInfo.IsResourceRequest {
+			klog.V(3).Infof("Not a resource request skipping proxying")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// Request has already been proxied once, it must be served locally
+		if r.Header.Get(PeerProxiedHeader) == "true" {
+			klog.V(3).Infof("Already rerouted once, skipping proxying to peer")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		// StorageVersion Informers and/or StorageVersionManager is not synced yet, pass request to next handler
+		// This will happen for self requests from the kube-apiserver because we have a poststarthook
+		// to ensure that external requests are not served until the StorageVersion Informer and
+		// StorageVersionManager has synced
+		if !h.HasFinishedSync() {
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		gvr := schema.GroupVersionResource{Group: requestInfo.APIGroup, Version: requestInfo.APIVersion, Resource: requestInfo.Resource}
+		if requestInfo.APIGroup == "" {
+			gvr.Group = "core"
+		}
+
+		// find servers that are capable of serving this request
+		serviceableByResp, err := h.findServiceableByServers(gvr, serverId)
+		if err != nil {
+			// this means that resource is an aggregated API or a CR since it wasn't found in SV informer cache, pass as it is
+			klog.Infof(fmt.Sprintf("no StorageVersion/APIServerID found for the GVR: %v skipping proxying", gvr))
+			handler.ServeHTTP(w, r)
+			return
+		}
+		// found the gvr locally, pass request to the next handler in local apiserver
+		if serviceableByResp.locallyServiceable {
+			klog.V(4).Infof("resource can be served locally, skipping proxying")
+			handler.ServeHTTP(w, r)
+			return
+		}
+
+		gv := schema.GroupVersion{Group: gvr.Group, Version: gvr.Version}
+
+		// if no apiservers were found that could serve the request, serve 404
+		if len(serviceableByResp.serviceableBy) == 0 {
+			klog.Errorf(fmt.Sprintf("GVR %v is not served by anything in this cluster", gvr))
+			responsewriters.ErrorNegotiated(apierrors.NewNotFound(schema.GroupResource{Group: gvr.Group, Resource: gvr.Resource},
+				fmt.Sprintf("%s.%s", gvr.Group, gvr.Resource)), s, gv, w, r)
+			return
+		}
+
+		// otherwise, randomly select an apiserver
+		rand := rand.Intn(len(serviceableByResp.serviceableBy))
+		apiserverId := serviceableByResp.serviceableBy[rand]
+
+		// finally proxy request to peer
+		ip, err := reconciler.GetEndpoint(apiserverId)
+		if err != nil {
+			klog.ErrorS(err, "error fetching endpoints from server leases")
+			responsewriters.ErrorNegotiated(apierrors.NewServiceUnavailable(fmt.Sprintf("Error proxying request for the requested GVR: %v, err: %v", gvr, err)), s, gv, w, r)
+		}
+
+		if ip == "" {
```

Comment:
Should peer selection be limited to routable peers?

Reply:
Hmm, how to get the list of routable peers? One thing we decided to do for beta was to use an egress dialer to check the proxy connection, maybe we can do this check there?

Reply:
Maybe what @benluddy is suggesting is to filter out peers to those with IPs before randomly picking one?

Reply:
> filter out peers to those with IPs before randomly picking one

Yes, that was what I meant to suggest. Thanks. Poor choice of words on my part.

Reply:
Got it, thanks for the suggestion. To confirm, this would mean finding endpoints (using etcd leases) for all the discovered apiservers first and then selecting one from them.

Reply:
I have updated the code to do this.